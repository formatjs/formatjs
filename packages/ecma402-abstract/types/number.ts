import {LDMLPluralRule} from './plural-rules'
import {LocaleData} from './core'

export type NumberFormatNotation =
  | 'standard'
  | 'scientific'
  | 'engineering'
  | 'compact'

export type RoundingPriorityType = 'auto' | 'morePrecision' | 'lessPrecision'

export type NumberFormatRoundingType =
  | 'morePrecision'
  | 'lessPrecision'
  | 'significantDigits'
  | 'fractionDigits'

export type RoundingModeType =
  | 'ceil'
  | 'floor'
  | 'expand'
  | 'trunc'
  | 'halfCeil'
  | 'halfFloor'
  | 'halfExpand'
  | 'halfTrunc'
  | 'halfEven'

export type UnsignedRoundingModeType =
  | 'infinity'
  | 'zero'
  | 'half-infinity'
  | 'half-zero'
  | 'half-even'

export type UseGroupingType = 'min2' | 'auto' | 'always' | boolean

export interface NumberFormatDigitOptions {
  minimumIntegerDigits?: number
  minimumSignificantDigits?: number
  maximumSignificantDigits?: number
  minimumFractionDigits?: number
  maximumFractionDigits?: number
  roundingPriority?: RoundingPriorityType
}

export interface NumberFormatDigitInternalSlots {
  minimumIntegerDigits: number
  minimumSignificantDigits?: number
  maximumSignificantDigits?: number
  roundingType: NumberFormatRoundingType
  // These two properties are only used when `roundingType` is "fractionDigits".
  minimumFractionDigits?: number
  maximumFractionDigits?: number
  notation?: NumberFormatNotation
  roundingIncrement?: number
  trailingZeroDisplay?: TrailingZeroDisplay
}

// All fields are optional due to de-duping
export type RawNumberLocaleData = LocaleData<NumberFormatLocaleInternalData>

export interface NumberFormatLocaleInternalData {
  units: UnitDataTable
  currencies: Record<string, CurrencyData>
  numbers: RawNumberData
  // Bc of relevantExtensionKeys in the spec
  nu: string[]
}

export interface UnitDataTable {
  simple: Record<string, UnitData>
  compound: Record<string, CompoundUnitData>
}

export interface UnitData {
  // A pattern where {0} is number placeholder. Example: "摂氏 {0} 度".
  long: LDMLPluralRuleMap<string>
  short: LDMLPluralRuleMap<string>
  narrow: LDMLPluralRuleMap<string>
  // perUnitPattern. See http://unicode.org/reports/tr35/tr35-general.html#perUnitPatterns
  perUnit: Record<'narrow' | 'short' | 'long', string | undefined>
}

// The values are patterns on how to compose simple units.
// For example, "{0} per {1}".
export interface CompoundUnitData {
  long: string
  short: string
  narrow: string
}

export interface CurrencyData {
  displayName: LDMLPluralRuleMap<string>
  symbol: string
  narrow: string
}

export type DecimalFormatNum =
  | '1000'
  | '10000'
  | '100000'
  | '1000000'
  | '10000000'
  | '100000000'
  | '1000000000'
  | '10000000000'
  | '100000000000'
  | '1000000000000'
  | '10000000000000'
  | '100000000000000'
export type NumberingSystem = string

/**
 * We only care about insertBetween bc we assume
 * `currencyMatch` & `surroundingMatch` are all the same
 *
 * @export
 * @interface CurrencySpacingData
 */
export interface CurrencySpacingData {
  beforeInsertBetween: string
  afterInsertBetween: string
}

export interface RawCurrencyData {
  currencySpacing: CurrencySpacingData
  standard: string
  accounting: string
  short?: Record<DecimalFormatNum, LDMLPluralRuleMap<string>>
  // IMPORTANT: We're making the assumption here that currency unitPattern
  // are the same for all LDMLPluralRule
  unitPattern: string
}

export interface SymbolsData {
  decimal: string
  group: string
  list: string
  percentSign: string
  plusSign: string
  minusSign: string
  exponential: string
  superscriptingExponent: string
  perMille: string
  infinity: string
  nan: string
  timeSeparator: string
  approximatelySign: string
}

export interface RawNumberData {
  nu: string[]
  // numberingSystem -> pattern
  symbols: Record<NumberingSystem, SymbolsData>
  // numberingSystem -> pattern
  decimal: Record<
    NumberingSystem,
    {
      // The standard number pattern of the decimal.
      standard: string
      // These two are compact notation mappings.
      long: Record<DecimalFormatNum, LDMLPluralRuleMap<string>>
      short: Record<DecimalFormatNum, LDMLPluralRuleMap<string>>
    }
  >
  percent: Record<NumberingSystem, string>
  currency: Record<NumberingSystem, RawCurrencyData>
}

export type LDMLPluralRuleMap<T> = Omit<
  Partial<Record<LDMLPluralRule, T>>,
  'other'
> & {
  other: T
}

export interface RawNumberFormatResult {
  formattedString: string
  roundedNumber: number
  integerDigitsCount: number
}

export type NumberFormatOptionsLocaleMatcher = 'lookup' | 'best fit'
export type NumberFormatOptionsStyle =
  | 'decimal'
  | 'percent'
  | 'currency'
  | 'unit'
export type NumberFormatOptionsCompactDisplay = 'short' | 'long'
export type NumberFormatOptionsCurrencyDisplay =
  | 'symbol'
  | 'code'
  | 'name'
  | 'narrowSymbol'
export type NumberFormatOptionsCurrencySign = 'standard' | 'accounting'
export type NumberFormatOptionsNotation = NumberFormatNotation
export type NumberFormatOptionsSignDisplay =
  | 'auto'
  | 'always'
  | 'never'
  | 'exceptZero'
  | 'negative'
export type NumberFormatOptionsUnitDisplay = 'long' | 'short' | 'narrow'
export type TrailingZeroDisplay = 'auto' | 'stripIfInteger'

export interface NumberFormatInternal extends NumberFormatDigitInternalSlots {
  locale: string
  dataLocale: string
  style: NumberFormatOptionsStyle
  currency?: string
  currencyDisplay: NumberFormatOptionsCurrencyDisplay
  unit?: string
  unitDisplay: NumberFormatOptionsUnitDisplay
  currencySign: NumberFormatOptionsCurrencySign
  notation: NumberFormatOptionsNotation
  compactDisplay: NumberFormatOptionsCompactDisplay
  signDisplay: NumberFormatOptionsSignDisplay
  useGrouping?: UseGroupingType
  pl: Intl.PluralRules
  boundFormat?: Intl.NumberFormat['format']
  numberingSystem: string
  // Locale-dependent formatter data
  dataLocaleData: NumberFormatLocaleInternalData
  roundingMode?: RoundingModeType
}

export type NumberFormatOptions = Omit<
  Intl.NumberFormatOptions,
  'signDisplay' | 'useGrouping'
> &
  NumberFormatDigitOptions & {
    localeMatcher?: NumberFormatOptionsLocaleMatcher
    style?: NumberFormatOptionsStyle
    compactDisplay?: NumberFormatOptionsCompactDisplay
    currencyDisplay?: NumberFormatOptionsCurrencyDisplay
    currencySign?: NumberFormatOptionsCurrencySign
    notation?: NumberFormatOptionsNotation
    signDisplay?: NumberFormatOptionsSignDisplay
    unit?: string
    unitDisplay?: NumberFormatOptionsUnitDisplay
    numberingSystem?: string
    trailingZeroDisplay?: TrailingZeroDisplay
    roundingPriority?: RoundingPriorityType
    roundingIncrement?: number
    roundingMode?: RoundingModeType
    useGrouping?: UseGroupingType
  }

export type ResolvedNumberFormatOptions = Intl.ResolvedNumberFormatOptions &
  Pick<
    NumberFormatInternal,
    | 'currencySign'
    | 'unit'
    | 'unitDisplay'
    | 'notation'
    | 'compactDisplay'
    | 'signDisplay'
  >

export type NumberFormatPartTypes =
  | Intl.NumberFormatPartTypes
  | 'exponentSeparator'
  | 'exponentMinusSign'
  | 'exponentInteger'
  | 'compact'
  | 'unit'
  | 'literal'
  | 'approximatelySign'

export interface NumberFormatPart {
  type: NumberFormatPartTypes
  value: string
  source?: string
}
