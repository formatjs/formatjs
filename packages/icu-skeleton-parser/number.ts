import type {NumberFormatOptions} from '@formatjs/ecma402-abstract'
import {WHITE_SPACE_REGEX} from './regex.generated'

export interface ExtendedNumberFormatOptions extends NumberFormatOptions {
  scale?: number
}
export interface NumberSkeletonToken {
  stem: string
  options: string[]
}

export function parseNumberSkeletonFromString(
  skeleton: string
): NumberSkeletonToken[] {
  if (skeleton.length === 0) {
    throw new Error('Number skeleton cannot be empty')
  }
  // Parse the skeleton
  const stringTokens = skeleton
    .split(WHITE_SPACE_REGEX)
    .filter(x => x.length > 0)

  const tokens: NumberSkeletonToken[] = []
  for (const stringToken of stringTokens) {
    let stemAndOptions = stringToken.split('/')
    if (stemAndOptions.length === 0) {
      throw new Error('Invalid number skeleton')
    }

    const [stem, ...options] = stemAndOptions
    for (const option of options) {
      if (option.length === 0) {
        throw new Error('Invalid number skeleton')
      }
    }

    tokens.push({stem, options})
  }
  return tokens
}

function icuUnitToEcma(unit: string): ExtendedNumberFormatOptions['unit'] {
  return unit.replace(/^(.*?)-/, '') as ExtendedNumberFormatOptions['unit']
}

const FRACTION_PRECISION_REGEX = /^\.(?:(0+)(\*)?|(#+)|(0+)(#+))$/g
const SIGNIFICANT_PRECISION_REGEX = /^(@+)?(\+|#+)?[rs]?$/g
const INTEGER_WIDTH_REGEX = /(\*)(0+)|(#+)(0+)|(0+)/g
const CONCISE_INTEGER_WIDTH_REGEX = /^(0+)$/

function parseSignificantPrecision(str: string): ExtendedNumberFormatOptions {
  const result: ExtendedNumberFormatOptions = {}
  if (str[str.length - 1] === 'r') {
    result.roundingPriority = 'morePrecision'
  } else if (str[str.length - 1] === 's') {
    result.roundingPriority = 'lessPrecision'
  }
  str.replace(
    SIGNIFICANT_PRECISION_REGEX,
    function (_: string, g1: string, g2: string | number) {
      // @@@ case
      if (typeof g2 !== 'string') {
        result.minimumSignificantDigits = g1.length
        result.maximumSignificantDigits = g1.length
      }
      // @@@+ case
      else if (g2 === '+') {
        result.minimumSignificantDigits = g1.length
      }
      // .### case
      else if (g1[0] === '#') {
        result.maximumSignificantDigits = g1.length
      }
      // .@@## or .@@@ case
      else {
        result.minimumSignificantDigits = g1.length
        result.maximumSignificantDigits =
          g1.length + (typeof g2 === 'string' ? g2.length : 0)
      }
      return ''
    }
  )
  return result
}

function parseSign(str: string): ExtendedNumberFormatOptions | undefined {
  switch (str) {
    case 'sign-auto':
      return {
        signDisplay: 'auto',
      }
    case 'sign-accounting':
    case '()':
      return {
        currencySign: 'accounting',
      }
    case 'sign-always':
    case '+!':
      return {
        signDisplay: 'always',
      }
    case 'sign-accounting-always':
    case '()!':
      return {
        signDisplay: 'always',
        currencySign: 'accounting',
      }
    case 'sign-except-zero':
    case '+?':
      return {
        signDisplay: 'exceptZero',
      }
    case 'sign-accounting-except-zero':
    case '()?':
      return {
        signDisplay: 'exceptZero',
        currencySign: 'accounting',
      }
    case 'sign-never':
    case '+_':
      return {
        signDisplay: 'never',
      }
  }
}

function parseConciseScientificAndEngineeringStem(
  stem: string
): ExtendedNumberFormatOptions | undefined {
  // Engineering
  let result: ExtendedNumberFormatOptions | undefined
  if (stem[0] === 'E' && stem[1] === 'E') {
    result = {
      notation: 'engineering',
    }
    stem = stem.slice(2)
  } else if (stem[0] === 'E') {
    result = {
      notation: 'scientific',
    }
    stem = stem.slice(1)
  }
  if (result) {
    const signDisplay = stem.slice(0, 2)
    if (signDisplay === '+!') {
      result.signDisplay = 'always'
      stem = stem.slice(2)
    } else if (signDisplay === '+?') {
      result.signDisplay = 'exceptZero'
      stem = stem.slice(2)
    }
    if (!CONCISE_INTEGER_WIDTH_REGEX.test(stem)) {
      throw new Error('Malformed concise eng/scientific notation')
    }
    result.minimumIntegerDigits = stem.length
  }
  return result
}

function parseNotationOptions(opt: string): ExtendedNumberFormatOptions {
  const result: ExtendedNumberFormatOptions = {}
  const signOpts = parseSign(opt)
  if (signOpts) {
    return signOpts
  }
  return result
}

/**
 * https://github.com/unicode-org/icu/blob/master/docs/userguide/format_parse/numbers/skeletons.md#skeleton-stems-and-options
 */
export function parseNumberSkeleton(
  tokens: NumberSkeletonToken[]
): ExtendedNumberFormatOptions {
  let result: ExtendedNumberFormatOptions = {}
  for (const token of tokens) {
    switch (token.stem) {
      case 'percent':
      case '%':
        result.style = 'percent'
        continue
      case '%x100':
        result.style = 'percent'
        result.scale = 100
        continue
      case 'currency':
        result.style = 'currency'
        result.currency = token.options[0]
        continue
      case 'group-off':
      case ',_':
        result.useGrouping = false
        continue
      case 'precision-integer':
      case '.':
        result.maximumFractionDigits = 0
        continue
      case 'measure-unit':
      case 'unit':
        result.style = 'unit'
        result.unit = icuUnitToEcma(token.options[0])
        continue
      case 'compact-short':
      case 'K':
        result.notation = 'compact'
        result.compactDisplay = 'short'
        continue
      case 'compact-long':
      case 'KK':
        result.notation = 'compact'
        result.compactDisplay = 'long'
        continue
      case 'scientific':
        result = {
          ...result,
          notation: 'scientific',
          ...token.options.reduce(
            (all, opt) => ({...all, ...parseNotationOptions(opt)}),
            {}
          ),
        }
        continue
      case 'engineering':
        result = {
          ...result,
          notation: 'engineering',
          ...token.options.reduce(
            (all, opt) => ({...all, ...parseNotationOptions(opt)}),
            {}
          ),
        }
        continue
      case 'notation-simple':
        result.notation = 'standard'
        continue
      // https://github.com/unicode-org/icu/blob/master/icu4c/source/i18n/unicode/unumberformatter.h
      case 'unit-width-narrow':
        result.currencyDisplay = 'narrowSymbol'
        result.unitDisplay = 'narrow'
        continue
      case 'unit-width-short':
        result.currencyDisplay = 'code'
        result.unitDisplay = 'short'
        continue
      case 'unit-width-full-name':
        result.currencyDisplay = 'name'
        result.unitDisplay = 'long'
        continue
      case 'unit-width-iso-code':
        result.currencyDisplay = 'symbol'
        continue
      case 'scale':
        result.scale = parseFloat(token.options[0])
        continue
      // https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#integer-width
      case 'integer-width':
        if (token.options.length > 1) {
          throw new RangeError(
            'integer-width stems only accept a single optional option'
          )
        }
        token.options[0].replace(
          INTEGER_WIDTH_REGEX,
          function (
            _: string,
            g1: string,
            g2: string,
            g3: string,
            g4: string,
            g5: string
          ) {
            if (g1) {
              result.minimumIntegerDigits = g2.length
            } else if (g3 && g4) {
              throw new Error(
                'We currently do not support maximum integer digits'
              )
            } else if (g5) {
              throw new Error(
                'We currently do not support exact integer digits'
              )
            }
            return ''
          }
        )
        continue
    }
    // https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#integer-width
    if (CONCISE_INTEGER_WIDTH_REGEX.test(token.stem)) {
      result.minimumIntegerDigits = token.stem.length
      continue
    }
    if (FRACTION_PRECISION_REGEX.test(token.stem)) {
      // Precision
      // https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#fraction-precision
      // precision-integer case
      if (token.options.length > 1) {
        throw new RangeError(
          'Fraction-precision stems only accept a single optional option'
        )
      }
      token.stem.replace(
        FRACTION_PRECISION_REGEX,
        function (
          _: string,
          g1: string,
          g2: string | number,
          g3: string,
          g4: string,
          g5: string
        ) {
          // .000* case (before ICU67 it was .000+)
          if (g2 === '*') {
            result.minimumFractionDigits = g1.length
          }
          // .### case
          else if (g3 && g3[0] === '#') {
            result.maximumFractionDigits = g3.length
          }
          // .00## case
          else if (g4 && g5) {
            result.minimumFractionDigits = g4.length
            result.maximumFractionDigits = g4.length + g5.length
          } else {
            result.minimumFractionDigits = g1.length
            result.maximumFractionDigits = g1.length
          }
          return ''
        }
      )

      const opt = token.options[0]
      // https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#trailing-zero-display
      if (opt === 'w') {
        result = {...result, trailingZeroDisplay: 'stripIfInteger'}
      } else if (opt) {
        result = {...result, ...parseSignificantPrecision(opt)}
      }
      continue
    }
    // https://unicode-org.github.io/icu/userguide/format_parse/numbers/skeletons.html#significant-digits-precision
    if (SIGNIFICANT_PRECISION_REGEX.test(token.stem)) {
      result = {...result, ...parseSignificantPrecision(token.stem)}
      continue
    }
    const signOpts = parseSign(token.stem)
    if (signOpts) {
      result = {...result, ...signOpts}
    }
    const conciseScientificAndEngineeringOpts =
      parseConciseScientificAndEngineeringStem(token.stem)
    if (conciseScientificAndEngineeringOpts) {
      result = {...result, ...conciseScientificAndEngineeringOpts}
    }
  }
  return result
}
