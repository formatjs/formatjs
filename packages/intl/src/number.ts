import {Formatters, IntlFormatters, CustomFormats, OnErrorFn} from './types'
import {getNamedFormat, filterProps} from './utils'
import {IntlError, IntlErrorCode} from './error'
import {NumberFormatOptions} from '@formatjs/ecma402-abstract'

const NUMBER_FORMAT_OPTIONS: Array<keyof NumberFormatOptions> = [
  'style',
  'currency',
  'currencyDisplay',
  'unit',
  'unitDisplay',
  'useGrouping',

  'minimumIntegerDigits',
  'minimumFractionDigits',
  'maximumFractionDigits',
  'minimumSignificantDigits',
  'maximumSignificantDigits',

  // ES2020 NumberFormat
  'compactDisplay',
  'currencyDisplay',
  'currencySign',
  'notation',
  'signDisplay',
  'unit',
  'unitDisplay',
  'numberingSystem',
]

export function getFormatter(
  {
    locale,
    formats,
    onError,
  }: {
    locale: string

    formats: CustomFormats
    onError: OnErrorFn
  },
  getNumberFormat: Formatters['getNumberFormat'],
  options: Parameters<IntlFormatters['formatNumber']>[1] = {}
): Intl.NumberFormat {
  const {format} = options
  const defaults = ((format &&
    getNamedFormat(formats!, 'number', format, onError)) ||
    {}) as NumberFormatOptions
  const filteredOptions = filterProps(
    options,
    NUMBER_FORMAT_OPTIONS,
    defaults
  ) as NumberFormatOptions

  return getNumberFormat(locale, filteredOptions)
}

export function formatNumber(
  config: {
    locale: string

    formats: CustomFormats
    onError: OnErrorFn
  },
  getNumberFormat: Formatters['getNumberFormat'],
  value: Parameters<IntlFormatters['formatNumber']>[0],
  options: Parameters<IntlFormatters['formatNumber']>[1] = {}
): string {
  try {
    return getFormatter(config, getNumberFormat, options).format(value)
  } catch (e) {
    config.onError(
      new IntlError(IntlErrorCode.FORMAT_ERROR, 'Error formatting number.', e)
    )
  }

  return String(value)
}

export function formatNumberToParts(
  config: {
    locale: string
    formats: CustomFormats
    onError: OnErrorFn
  },
  getNumberFormat: Formatters['getNumberFormat'],
  value: Parameters<IntlFormatters['formatNumber']>[0],
  options: Parameters<IntlFormatters['formatNumber']>[1] = {}
): Intl.NumberFormatPart[] {
  try {
    return getFormatter(config, getNumberFormat, options).formatToParts(
      value as number
    )
  } catch (e) {
    config.onError(
      new IntlError(IntlErrorCode.FORMAT_ERROR, 'Error formatting number.', e)
    )
  }

  return []
}
