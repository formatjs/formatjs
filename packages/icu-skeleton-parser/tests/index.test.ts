import {parseDateTimeSkeleton, parseNumberSkeleton} from '..'
import {parseNumberSkeletonFromString} from '../number'

test.each([
  `yyyy.MM.dd G 'at' HH:mm:ss zzzz`,
  `EEE, MMM d, ''yy`,
  `h:mm a`,
  ``,
])('case: %p', skeleton => {
  expect(parseDateTimeSkeleton(skeleton)).toMatchSnapshot()
})

test.each([
  'percent .##',
  '.##',
  '.##/w',
  '.',
  '% .##',
  '.##/@##r',
  '.##/@##s',
  'percent .000*',
  'percent .0###',
  'percent .00/@##',
  'percent .00/@@@',
  'percent .00/@@@@*',
  'percent scale/0.01',
  'currency/CAD .',
  'currency/GBP .0*/@@@',
  'currency/GBP .00##/@@@',
  'currency/GBP .00##/@@@ unit-width-full-name',
  'measure-unit/length-meter .00##/@@@',
  'measure-unit/length-meter .00##/@@@ unit-width-full-name',
  'compact-short',
  'compact-long',
  'scientific',
  'scientific/sign-always',
  'scientific/+ee/sign-always',
  'engineering',
  'engineering/sign-except-zero',
  'notation-simple',
  'sign-auto',
  'sign-always',
  '+!',
  'sign-never',
  '+_',
  'sign-accounting',
  '()',
  'sign-accounting-always',
  '()!',
  'sign-except-zero',
  '+?',
  'sign-accounting-except-zero',
  '()?',
  '000',
  'integer-width/*000',
  'E0',
  'E+!00',
  'EE+?000',
  '%x100',
])('[parseNumberSkeleton] case: %p', skeleton => {
  expect(
    parseNumberSkeleton(parseNumberSkeletonFromString(skeleton))
  ).toMatchSnapshot()
})
