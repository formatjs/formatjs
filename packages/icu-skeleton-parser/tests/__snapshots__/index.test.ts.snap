// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[parseNumberSkeleton] case: "% .##" 1`] = `
{
  "maximumFractionDigits": 2,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "%x100" 1`] = `
{
  "scale": 100,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "()!" 1`] = `
{
  "currencySign": "accounting",
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "()" 1`] = `
{
  "currencySign": "accounting",
}
`;

exports[`[parseNumberSkeleton] case: "()?" 1`] = `
{
  "currencySign": "accounting",
  "signDisplay": "exceptZero",
}
`;

exports[`[parseNumberSkeleton] case: "+!" 1`] = `
{
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "+?" 1`] = `
{
  "signDisplay": "exceptZero",
}
`;

exports[`[parseNumberSkeleton] case: "+_" 1`] = `
{
  "signDisplay": "never",
}
`;

exports[`[parseNumberSkeleton] case: "." 1`] = `
{
  "maximumFractionDigits": 0,
}
`;

exports[`[parseNumberSkeleton] case: ".##" 1`] = `
{
  "maximumFractionDigits": 2,
}
`;

exports[`[parseNumberSkeleton] case: ".##/@##r" 1`] = `
{
  "maximumFractionDigits": 2,
  "maximumSignificantDigits": 3,
  "minimumSignificantDigits": 1,
  "roundingPriority": "morePrecision",
}
`;

exports[`[parseNumberSkeleton] case: ".##/@##s" 1`] = `
{
  "maximumFractionDigits": 2,
  "maximumSignificantDigits": 3,
  "minimumSignificantDigits": 1,
  "roundingPriority": "lessPrecision",
}
`;

exports[`[parseNumberSkeleton] case: ".##/w" 1`] = `
{
  "maximumFractionDigits": 2,
  "trailingZeroDisplay": "stripIfInteger",
}
`;

exports[`[parseNumberSkeleton] case: ".00/w currency/CAD" 1`] = `
{
  "currency": "CAD",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "style": "currency",
  "trailingZeroDisplay": "stripIfInteger",
}
`;

exports[`[parseNumberSkeleton] case: "000" 1`] = `
{
  "minimumIntegerDigits": 3,
}
`;

exports[`[parseNumberSkeleton] case: "E+!00" 1`] = `
{
  "minimumIntegerDigits": 2,
  "notation": "scientific",
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "E0" 1`] = `
{
  "minimumIntegerDigits": 1,
  "notation": "scientific",
}
`;

exports[`[parseNumberSkeleton] case: "EE+?000" 1`] = `
{
  "minimumIntegerDigits": 3,
  "notation": "engineering",
  "signDisplay": "exceptZero",
}
`;

exports[`[parseNumberSkeleton] case: "compact-long" 1`] = `
{
  "compactDisplay": "long",
  "notation": "compact",
}
`;

exports[`[parseNumberSkeleton] case: "compact-short" 1`] = `
{
  "compactDisplay": "short",
  "notation": "compact",
}
`;

exports[`[parseNumberSkeleton] case: "currency/CAD ." 1`] = `
{
  "currency": "CAD",
  "maximumFractionDigits": 0,
  "style": "currency",
}
`;

exports[`[parseNumberSkeleton] case: "currency/GBP .0*/@@@" 1`] = `
{
  "currency": "GBP",
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 1,
  "minimumSignificantDigits": 3,
  "style": "currency",
}
`;

exports[`[parseNumberSkeleton] case: "currency/GBP .00##/@@@ unit-width-full-name" 1`] = `
{
  "currency": "GBP",
  "currencyDisplay": "name",
  "maximumFractionDigits": 4,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "currency",
  "unitDisplay": "long",
}
`;

exports[`[parseNumberSkeleton] case: "currency/GBP .00##/@@@" 1`] = `
{
  "currency": "GBP",
  "maximumFractionDigits": 4,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "currency",
}
`;

exports[`[parseNumberSkeleton] case: "engineering" 1`] = `
{
  "notation": "engineering",
}
`;

exports[`[parseNumberSkeleton] case: "engineering/sign-except-zero" 1`] = `
{
  "notation": "engineering",
  "signDisplay": "exceptZero",
}
`;

exports[`[parseNumberSkeleton] case: "integer-width/*000" 1`] = `
{
  "minimumIntegerDigits": 3,
}
`;

exports[`[parseNumberSkeleton] case: "measure-unit/length-meter .00##/@@@ unit-width-full-name" 1`] = `
{
  "currencyDisplay": "name",
  "maximumFractionDigits": 4,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "unit",
  "unit": "meter",
  "unitDisplay": "long",
}
`;

exports[`[parseNumberSkeleton] case: "measure-unit/length-meter .00##/@@@" 1`] = `
{
  "maximumFractionDigits": 4,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "unit",
  "unit": "meter",
}
`;

exports[`[parseNumberSkeleton] case: "notation-simple" 1`] = `
{
  "notation": "standard",
}
`;

exports[`[parseNumberSkeleton] case: "percent .##" 1`] = `
{
  "maximumFractionDigits": 2,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "percent .0###" 1`] = `
{
  "maximumFractionDigits": 4,
  "minimumFractionDigits": 1,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "percent .00/@##" 1`] = `
{
  "maximumFractionDigits": 2,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 1,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "percent .00/@@@" 1`] = `
{
  "maximumFractionDigits": 2,
  "maximumSignificantDigits": 3,
  "minimumFractionDigits": 2,
  "minimumSignificantDigits": 3,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "percent .00/@@@@*" 1`] = `
{
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "percent .000*" 1`] = `
{
  "minimumFractionDigits": 3,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "percent scale/0.01" 1`] = `
{
  "scale": 0.01,
  "style": "percent",
}
`;

exports[`[parseNumberSkeleton] case: "scientific" 1`] = `
{
  "notation": "scientific",
}
`;

exports[`[parseNumberSkeleton] case: "scientific/+ee/sign-always" 1`] = `
{
  "notation": "scientific",
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "scientific/sign-always" 1`] = `
{
  "notation": "scientific",
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "sign-accounting" 1`] = `
{
  "currencySign": "accounting",
}
`;

exports[`[parseNumberSkeleton] case: "sign-accounting-always" 1`] = `
{
  "currencySign": "accounting",
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "sign-accounting-except-zero" 1`] = `
{
  "currencySign": "accounting",
  "signDisplay": "exceptZero",
}
`;

exports[`[parseNumberSkeleton] case: "sign-always" 1`] = `
{
  "signDisplay": "always",
}
`;

exports[`[parseNumberSkeleton] case: "sign-auto" 1`] = `
{
  "signDisplay": "auto",
}
`;

exports[`[parseNumberSkeleton] case: "sign-except-zero" 1`] = `
{
  "signDisplay": "exceptZero",
}
`;

exports[`[parseNumberSkeleton] case: "sign-never" 1`] = `
{
  "signDisplay": "never",
}
`;

exports[`case: "" 1`] = `{}`;

exports[`case: "EEE, MMM d, ''yy" 1`] = `
{
  "day": "numeric",
  "month": "short",
  "weekday": "short",
  "year": "2-digit",
}
`;

exports[`case: "h:mm a" 1`] = `
{
  "hour": "numeric",
  "hour12": true,
  "hourCycle": "h12",
  "minute": "2-digit",
}
`;

exports[`case: "yyyy.MM.dd G 'at' HH:mm:ss zzzz" 1`] = `
{
  "day": "2-digit",
  "era": "short",
  "hour": "2-digit",
  "hourCycle": "h23",
  "minute": "2-digit",
  "month": "2-digit",
  "second": "2-digit",
  "timeZoneName": "long",
  "year": "numeric",
}
`;
